compiler/kernel/bufferfilter.go

```diff
@@ -73,41 +80,47 @@ func CompileBufferFilter(e dag.Expr) (*expr.BufferFilter, error) {
 	}
 }
 
-func isFieldEqualOrIn(e *dag.BinaryExpr) (*astzed.Primitive, string) {
+// XXX isFieldEqualOrIn should work for any paths not just top-level fields.
+// See issue #3412
+
+func isFieldEqualOrIn(zctx *zed.Context, e *dag.BinaryExpr) (*zed.Value, string) {
 	if dag.IsTopLevelField(e.LHS) && e.Op == "=" {
-		if literal, ok := e.RHS.(*astzed.Primitive); ok {
-			return literal, "="
+		if literal, ok := e.RHS.(*dag.Literal); ok {
+			val, err := zson.ParseValue(zctx, literal.Value)
+			if err != nil {
+				return nil, ""
+			}
+			return &val, "="
 		}
 	} else if dag.IsTopLevelField(e.RHS) && e.Op == "in" {
-		if literal, ok := e.LHS.(*astzed.Primitive); ok && literal.Type != "net" {
-			return literal, "in"
+		if literal, ok := e.LHS.(*dag.Literal); ok {
+			val, err := zson.ParseValue(zctx, literal.Value)
+			if err != nil {
+				return nil, ""
+			}
```

Comment:
We should probably change the signature of `isFieldEqualOrIn` to return these `zson.ParseValue` errors rather than swallowing them.